sympy/combinatorics/rewritingsystem.py

```diff
@@ -285,3 +289,236 @@ def reduce(self, word, exclude=None):
+                fsm.add_state(rule, is_dead=True)
+
+        # Add accept states.
+        for r in all_rules:
+            prop_prefix = proper_prefixes[r]
+            for elem in prop_prefix:
+                if not elem in fsm.states:
+                    fsm.add_state(elem, is_accept=True)
+
+        # Add transitions for every state
+        for state in fsm.states:
+            current_state_name = state
+            current_state_type = fsm.states[state].state_type
+            if current_state_type == "start":
+                for letter in automaton_alphabet:
+                    if letter in fsm.states:
+                        fsm.states[state].add_transition(letter, letter)
+                    else:
+                        fsm.states[state].add_transition(letter, current_state_name)
+            elif current_state_type == "accept":
+                for letter in automaton_alphabet:
+                    next = current_state_name*letter
+                    len_next_word = len(next)
+                    while True:
+                        if len(next) <= 1:
+                            if next in fsm.states:
+                                fsm.states[state].add_transition(letter, next)
+                            else:
+                                fsm.states[state].add_transition(letter, 'start')
+                            break
+                        else:
+                            if next in fsm.states:
+                                fsm.states[state].add_transition(letter, next)
+                                break
+                            next = next.subword(1, len(next))
+        return fsm
+
+    def add_to_automaton(self, rules):
+        # Automaton vairables
+        automaton_alphabet = []
+        left_hand_rules = []
+        proper_prefixes = {}
+
+        # compute the inverses of all the new rules added
+        all_rules = rules
+        inverse_rules = self.compute_inverse_rules(all_rules)
+        all_rules.update(inverse_rules)
+
+        # All the transition symbols in the automaton
+        generators = list(self.alphabet)
+        generators += [gen**-1 for gen in generators]
+
+        # The symbols present in the new rules are the symbols to be verified at each state.
+        # computes the automaton_alphabet, as the transitions solely depend upon the new states.
+        for rule in rules:
+            automaton_alphabet += rule.letter_form_elm
+        automaton_alphabet = set(automaton_alphabet)
+
+        # Store the complete left hand side of the rules - dead states.
+        left_hand_rules = list(all_rules)
+
+        # Compute the proper prefixes for every rule.
+        for r in all_rules:
+            proper_prefixes[r] = []
+            letter_word_array = [s for s in r.letter_form_elm]
+            for i in range (1, len(letter_word_array)):
+                letter_word_array[i] = letter_word_array[i-1]*letter_word_array[i]
+            proper_prefixes[r] = letter_word_array
```

Comment:
Hmm, this and pretty much the rest of this function is almost an exact copy of `construct_automaton`. Could you just call `add_to_automaton(self.rules)` there instead of having all of this code repeated?

Reply:
Have you considered this? I think it would work quite well and the code would be more compact.

Reply:
I've updated this